From 77b6dc79513fb03b04c0bf793114f1f9947a6e70 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Wed, 15 Aug 2018 10:45:27 +0800
Subject: [PATCH 4608/5242] MLK-19252-2 drm/bridge: adv7511: realize
 mode_fixup for lanes fixup

commit  54bd8256b30496a26d53b6b835c9b87e3d034182 from
https://source.codeaurora.org/external/imx/linux-imx.git

In the current implementation, the data lanes fixup is done in
the Bridge's mode_set() function which is too late to give the
upper layer Bridge a chance to check this change to be supported
or not and a chance to refuse this commit if it does not support
the requested data lanes number.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit c97f9febd7e42d7865921e2bb56648ebb13c0761)
(cherry picked from commit 36f25495a5c30e66e033cfd3d46d4693f50fc8f7)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/bridge/adv7511/adv7511.h     |    7 +++++++
 drivers/gpu/drm/bridge/adv7511/adv7511_drv.c |   13 +++++++++++++
 drivers/gpu/drm/bridge/adv7511/adv7533.c     |   16 +++++++++++++---
 3 files changed, 33 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/bridge/adv7511/adv7511.h b/drivers/gpu/drm/bridge/adv7511/adv7511.h
index 1b9a4f9..0b93106 100644
--- a/drivers/gpu/drm/bridge/adv7511/adv7511.h
+++ b/drivers/gpu/drm/bridge/adv7511/adv7511.h
@@ -402,6 +402,7 @@ static inline int adv7511_cec_init(struct device *dev, struct adv7511 *adv7511)
 void adv7533_dsi_power_on(struct adv7511 *adv);
 void adv7533_dsi_power_off(struct adv7511 *adv);
 void adv7533_mode_set(struct adv7511 *adv, struct drm_display_mode *mode);
+bool adv7533_mode_fixup(struct adv7511 *adv, struct drm_display_mode *mode);
 int adv7533_patch_registers(struct adv7511 *adv);
 int adv7533_patch_cec_registers(struct adv7511 *adv);
 int adv7533_attach_dsi(struct adv7511 *adv);
@@ -421,6 +422,12 @@ static inline void adv7533_mode_set(struct adv7511 *adv,
 {
 }
 
+static inline bool adv7533_mode_fixup(struct adv7511 *adv,
+				      struct drm_display_mode *mode)
+{
+	return true;
+}
+
 static inline int adv7533_patch_registers(struct adv7511 *adv)
 {
 	return -ENODEV;
diff --git a/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c b/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c
index 91e3226..a106883 100644
--- a/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c
+++ b/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c
@@ -863,6 +863,18 @@ static void adv7511_bridge_mode_set(struct drm_bridge *bridge,
 	adv7511_mode_set(adv, mode, adj_mode);
 }
 
+static bool adv7511_bridge_mode_fixup(struct drm_bridge *bridge,
+				      const struct drm_display_mode *mode,
+				      struct drm_display_mode *adjusted_mode)
+{
+	struct adv7511 *adv = bridge_to_adv7511(bridge);
+
+	if (adv->type == ADV7533 || adv->type == ADV7535)
+		return adv7533_mode_fixup(adv, adjusted_mode);
+
+	return true;
+}
+
 static int adv7511_bridge_attach(struct drm_bridge *bridge)
 {
 	struct adv7511 *adv = bridge_to_adv7511(bridge);
@@ -904,6 +916,7 @@ static int adv7511_bridge_attach(struct drm_bridge *bridge)
 	.enable = adv7511_bridge_enable,
 	.disable = adv7511_bridge_disable,
 	.mode_set = adv7511_bridge_mode_set,
+	.mode_fixup = adv7511_bridge_mode_fixup,
 	.attach = adv7511_bridge_attach,
 };
 
diff --git a/drivers/gpu/drm/bridge/adv7511/adv7533.c b/drivers/gpu/drm/bridge/adv7511/adv7533.c
index eb0e9b9..84851e4 100644
--- a/drivers/gpu/drm/bridge/adv7511/adv7533.c
+++ b/drivers/gpu/drm/bridge/adv7511/adv7533.c
@@ -110,11 +110,16 @@ void adv7533_dsi_power_off(struct adv7511 *adv)
 
 void adv7533_mode_set(struct adv7511 *adv, struct drm_display_mode *mode)
 {
+}
+
+bool adv7533_mode_fixup(struct adv7511 *adv,
+			struct drm_display_mode *mode)
+{
 	struct mipi_dsi_device *dsi = adv->dsi;
 	int lanes, ret;
 
 	if (adv->num_dsi_lanes != 4)
-		return;
+		return true;
 
 	if (mode->clock > 80000)
 		lanes = 4;
@@ -123,11 +128,16 @@ void adv7533_mode_set(struct adv7511 *adv, struct drm_display_mode *mode)
 
 	if (lanes != dsi->lanes) {
 		mipi_dsi_detach(dsi);
-		dsi->lanes = lanes;
+		swap(dsi->lanes, lanes);
 		ret = mipi_dsi_attach(dsi);
-		if (ret)
+		if (ret) {
 			dev_err(&dsi->dev, "failed to change host lanes\n");
+			swap(dsi->lanes, lanes);
+			return false;
+		}
 	}
+
+	return true;
 }
 
 int adv7533_patch_registers(struct adv7511 *adv)
-- 
1.7.9.5

